File: 01-游戏开发/01-服务端/游戏服务端架构/跨服RPC/go/main.go
// 跨服 RPC(Go 版):req_id 关联/超时/错误传播/单向推送。
// 帧:[u32 BE 帧长][u8 类型][u32 LE req_id][u8 路径长][路径][载荷];
// 类型 1=REQ 2=RESP(载荷前 1 字节状态码) 3=PUSH(req_id 恒 0)。
package main

import (
	"encoding/binary"
	"fmt"
	"os"
	"sort"
)

const (
	msgReq  = 1
	msgResp = 2
	msgPush = 3

	statusOK          = 0
	statusTimeout     = 1
	statusUnknownPath = 2
)

type frame struct {
	msgType uint8
	reqID   uint32
	path    string
	payload []byte
	status  uint8
}

func encodeFrame(f frame) []byte {
	p := []byte(f.path)
	n := 1 + 4 + 1 + len(p) + len(f.payload)
	if f.msgType == msgResp {
		n++
	}
	out := make([]byte, 4, 4+n)
	binary.BigEndian.PutUint32(out[0:], uint32(n)) // 大端长度前缀,同 gRPC
	out = append(out, f.msgType, byte(f.reqID), byte(f.reqID>>8),
		byte(f.reqID>>16), byte(f.reqID>>24), uint8(len(p)))
	out = append(out, p...)
	if f.msgType == msgResp {
		out = append(out, f.status)
	}
	return append(out, f.payload...)
}

func decodeFrame(data []byte) (frame, error) {
	if len(data) < 4 {
		return frame{}, fmt.Errorf("缺长度前缀")
	}
	n := int(binary.BigEndian.Uint32(data[0:4]))
	if len(data)-4 != n {
		return frame{}, fmt.Errorf("帧长 %d != 实际 %d", n, len(data)-4)
	}
	pos := 4
	msgType := data[pos]
	reqID := binary.LittleEndian.Uint32(data[pos+1:])
	plen := int(data[pos+5])
	pos += 6
	if msgType != msgReq && msgType != msgResp && msgType != msgPush {
		return frame{}, fmt.Errorf("未知消息类型 %d", msgType)
	}
	if pos+plen > len(data) {
		return frame{}, fmt.Errorf("路径截断")
	}
	f := frame{msgType: msgType, reqID: reqID, path: string(data[pos : pos+plen])}
	pos += plen
	if msgType == msgResp {
		f.status = data[pos]
		pos++
	}
	f.payload = data[pos:]
	return f, nil
}

// ---------------- 客户端:req_id 关联 + 超时 + 晚到丢弃 ----------------

type pending struct {
	path     string
	deadline int
	done     bool
	result   []byte
	errCode  uint8
}

type rpcClient struct {
	out     chan []byte
	pending map[uint32]*pending
	nextID  uint32
	pushes  []frame
}

func newClient(out chan []byte) *rpcClient {
	return &rpcClient{out: out, pending: map[uint32]*pending{}}
}

func (c *rpcClient) call(path string, timeoutMs int) uint32 {
	c.nextID++
	c.pending[c.nextID] = &pending{path: path, deadline: timeoutMs}
	c.out <- encodeFrame(frame{msgType: msgReq, reqID: c.nextID, path: path})
	return c.nextID
}

func (c *rpcClient) onFrame(raw []byte) string {
	f, err := decodeFrame(raw)
	if err != nil {
		panic(err)
	}
	if f.msgType == msgPush {
		c.pushes = append(c.pushes, f)
		return "push"
	}
	e, ok := c.pending[f.reqID]
	if !ok || e.done { // 超时后晚到的响应:静默丢弃
		return "dropped"
	}
	e.done = true
	if f.status == statusOK {
		e.result = f.payload
	} else {
		e.errCode = f.status
	}
	return "resolved"
}

func (c *rpcClient) expire(now int) {
	for _, e := range c.pending {
		if !e.done && now >= e.deadline {
			e.done, e.errCode = true, statusTimeout
		}
	}
}

func (c *rpcClient) drain(in chan []byte) {
	for {
		select {
		case raw := <-in:
			c.onFrame(raw)
			continue
		default:
		}
		return
	}
}

// ---------------- 服务端:handler 表 + 模拟延迟 ----------------

type handler struct {
	fn      func([]byte) ([]byte, uint8)
	delayMs int
}

func okFn(fn func([]byte) []byte) func([]byte) ([]byte, uint8) {
	return func(p []byte) ([]byte, uint8) { return fn(p), statusOK }
}

type scheduled struct {
	due     int
	reqID   uint32
	status  uint8
	payload []byte
}

type rpcServer struct {
	out      chan []byte
	handlers map[string]handler
	sched    []scheduled
}

func newServer(out chan []byte) *rpcServer {
	return &rpcServer{out: out, handlers: map[string]handler{}}
}

func (s *rpcServer) register(path string, fn func([]byte) ([]byte, uint8), delayMs int) {
	s.handlers[path] = handler{fn: fn, delayMs: delayMs}
}

func (s *rpcServer) pump(in chan []byte, now int) {
	for {
		select {
		case raw := <-in:
			f, err := decodeFrame(raw)
			if err != nil || f.msgType != msgReq {
				panic("服务端只处理 REQ")
			}
			h, known := s.handlers[f.path]
			if !known {
				s.sched = append(s.sched, scheduled{now, f.reqID, statusUnknownPath, nil})
				continue
			}
			payload, code := h.fn(f.payload)
			s.sched = append(s.sched, scheduled{now + h.delayMs, f.reqID, code, payload})
		default:
		}
		sort.SliceStable(s.sched, func(i, j int) bool { return s.sched[i].due < s.sched[j].due })
		s.emit(now)
		return
	}
}

func (s *rpcServer) emit(now int) {
	for len(s.sched) > 0 && s.sched[0].due <= now {
		sc := s.sched[0]
		s.sched = s.sched[1:]
		s.out <- encodeFrame(frame{msgType: msgResp, reqID: sc.reqID,
			payload: sc.payload, status: sc.status})
	}
}

func (s *rpcServer) push(path string, payload []byte) {
	s.out <- encodeFrame(frame{msgType: msgPush, path: path, payload: payload})
}

var failures int

func check(label string, cond bool) {
	if !cond {
		failures++
		fmt.Printf("[FAIL] %s\n", label)
		return
	}
	fmt.Printf("[ok] %s\n", label)
}

func takeOne(in chan []byte) []byte {
	select {
	case raw := <-in:
		return raw
	default:
	}
	return nil
}

func main() {
	// ---- 1. 帧编解码 ----
	raw := encodeFrame(frame{msgType: msgResp, reqID: 42,
		path: "/ZoneService/KickPlayer", payload: []byte("bye")})
	d, err := decodeFrame(raw)
	check("帧往返:类型/req_id/路径/载荷",
		err == nil && d.msgType == msgResp && d.reqID == 42
			&& d.path == "/ZoneService/KickPlayer" && string(d.payload) == "bye")
	check("长度前缀为 4 字节大端",
		binary.BigEndian.Uint32(raw[:4]) == uint32(len(raw)-4))
	if _, err = decodeFrame(raw[:len(raw)-1]); err == nil {
		check("帧长不符应报错", false)
	} else {
		check("帧长不符报错", true)
	}
	// ---- 2. 乱序响应按 req_id 正确关联 ----
	zoneToGlobal := make(chan []byte, 16) // zone -> global(请求方向)
	globalToZone := make(chan []byte, 16) // global -> zone(响应/推送方向)
	cli, srv := newClient(zoneToGlobal), newServer(globalToZone)
	srv.register("/Global/QueryRank", okFn(func([]byte) []byte { return []byte("rank-1") }), 30)
	srv.register("/Global/QueryMail", okFn(func([]byte) []byte { return []byte("mail-7") }), 5)
	srv.register("/Global/Ping", okFn(func([]byte) []byte { return []byte("pong") }), 0)

	a := cli.call("/Global/QueryRank", 200) // 慢
	b := cli.call("/Global/QueryMail", 200)
	c := cli.call("/Global/Ping", 200)
	srv.pump(zoneToGlobal, 0)      // 入队三响应,emit 只放行 due=0 的 Ping
	cli.onFrame(takeOne(globalToZone))
	check("乱序到达仍正确关联(Ping 最先回)",
		string(cli.pending[c].result) == "pong" && cli.pending[a].result == nil)
	srv.emit(30) // 30ms 时 mail(5) 与 rank(30) 都到期
	cli.drain(globalToZone)
	check("三个并发调用全部配对成功",
		string(cli.pending[a].result) == "rank-1"
			&& string(cli.pending[b].result) == "mail-7"
			&& string(cli.pending[c].result) == "pong")
	// ---- 3. 超时 + 晚到响应丢弃 ----
	srv.register("/Global/SlowEcho", okFn(func(p []byte) []byte { return p }), 200)
	d2 := cli.call("/Global/SlowEcho", 100)
	srv.pump(zoneToGlobal, 0)
	cli.expire(110)
	check("110ms 判 TIMEOUT", cli.pending[d2].errCode == statusTimeout)
	srv.emit(200) // 200ms 响应才到
	late := cli.onFrame(takeOne(globalToZone))
	check("晚到响应被静默丢弃", late == "dropped")
	// ---- 4. 错误传播(状态码)与未知路径 ----
	srv.register("/Global/Boom", func([]byte) ([]byte, uint8) { return nil, 7 }, 0)
	e := cli.call("/Global/Boom", 100)
	u := cli.call("/Global/NoSuch", 100)
	srv.pump(zoneToGlobal, 0)
	srv.emit(0)
	cli.drain(globalToZone)
	check("业务错误按状态码 7 传播", cli.pending[e].errCode == 7)
	check("未知路径返回状态码 2", cli.pending[u].errCode == statusUnknownPath)
	// ---- 5. 单向推送 ----
	srv.push("/ZoneService/KickPlayer", []byte("uid=1001"))
	cli.onFrame(takeOne(globalToZone))
	check("PUSH 无配对直达推送队列",
		len(cli.pushes) == 1 && cli.pushes[0].path == "/ZoneService/KickPlayer"
			&& string(cli.pushes[0].payload) == "uid=1001")
	if failures > 0 {
		fmt.Printf("\n%d 项断言失败\n", failures)
		os.Exit(1)
	}
	fmt.Println("\n全部断言通过")
}

File: 01-游戏开发/01-服务端/游戏服务端架构/跨服RPC/python/main.py
# -*- coding: utf-8 -*-
"""跨服 RPC:一根有序连接上的请求号关联、超时、错误传播与单向推送。

帧格式(对照 gRPC Length-Prefixed-Message 的大端 4 字节长度前缀):
  [u32 BE 帧长][u8 类型][u32 BE req_id][u8 路径长][路径][载荷]
  类型 1=REQ 2=RESP(载荷前有 1 字节状态码) 3=PUSH(req_id 恒 0)
"""
import struct
from collections import deque

MSG_REQ, MSG_RESP, MSG_PUSH = 1, 2, 3
STATUS_OK, STATUS_TIMEOUT, STATUS_UNKNOWN_PATH = 0, 1, 2


class FrameError(ValueError):
    pass


class Frame:
    __slots__ = ("msg_type", "req_id", "path", "payload", "status")

    def __init__(self, msg_type, req_id, path, payload=b"", status=0):
        self.msg_type, self.req_id = msg_type, req_id
        self.path, self.payload, self.status = path, payload, status


def encode_frame(f):
    p = f.path.encode()
    body = struct.pack("<BIB", f.msg_type, f.req_id, len(p)) + p
    if f.msg_type == MSG_RESP:
        body += struct.pack("<B", f.status)
    body += f.payload
    return struct.pack(">I", len(body)) + body


def decode_frame(data):
    if len(data) < 4:
        raise FrameError("缺长度前缀")
    n = struct.unpack(">I", data[:4])[0]
    if len(data) - 4 != n:
        raise FrameError("帧长 %d != 实际 %d" % (n, len(data) - 4))
    pos = 4
    msg_type, req_id, plen = struct.unpack_from("<BIB", data, pos)
    pos += 6
    if msg_type not in (MSG_REQ, MSG_RESP, MSG_PUSH):
        raise FrameError("未知消息类型 %d" % msg_type)
    if pos + plen > len(data):
        raise FrameError("路径截断")
    path = data[pos:pos + plen].decode()
    pos += plen
    status = 0
    if msg_type == MSG_RESP:
        status = data[pos]
        pos += 1
    return Frame(msg_type, req_id, path, data[pos:], status)


# ---------------- 客户端:req_id 关联 + 超时 + 晚到丢弃 ----------------

class Pending:
    __slots__ = ("path", "deadline", "done", "result", "error")

    def __init__(self, path, deadline):
        self.path, self.deadline = path, deadline
        self.done, self.result, self.error = False, None, None


class RpcClient:
    def __init__(self, out_q):
        self.out = out_q
        self.pending = {}      # req_id -> Pending
        self.next_id = 1
        self.pushes = []       # 单向推送落地处

    def call(self, path, payload=b"", timeout_ms=100):
        rid = self.next_id
        self.next_id += 1
        self.pending[rid] = Pending(path, timeout_ms)
        self.out.append(encode_frame(Frame(MSG_REQ, rid, path, payload)))
        return rid

    def on_frame(self, raw):
        f = decode_frame(raw)
        if f.msg_type == MSG_PUSH:
            if f.req_id != 0:
                raise FrameError("PUSH 的 req_id 必须为 0")
            self.pushes.append(f)
            return
        entry = self.pending.get(f.req_id)
        if entry is None or entry.done:      # 超时后晚到的响应:静默丢弃
            return "dropped"
        entry.done = True
        if f.status == STATUS_OK:
            entry.result = f.payload
        else:
            entry.error = f.status
        return "resolved"

    def expire(self, now):
        """超时本地判定;判完的 pending 保留 done 态以丢弃晚到响应。"""
        for e in self.pending.values():
            if not e.done and now >= e.deadline:
                e.done, e.error = True, STATUS_TIMEOUT


# ---------------- 服务端:handler 表 + 模拟延迟 ----------------

class RpcServer:
    def __init__(self, out_q):
        self.out = out_q
        self.handlers = {}    # path -> (fn, delay_ms)
        self.scheduled = []   # (due_ms, req_id, path, status, payload)

    def register(self, path, fn, delay_ms=0):
        self.handlers[path] = (fn, delay_ms)

    def push(self, path, payload=b""):
        self.out.append(encode_frame(Frame(MSG_PUSH, 0, path, payload)))

    def pump(self, in_q, now):
        while in_q:
            f = decode_frame(in_q.popleft())
            if f.msg_type != MSG_REQ:
                raise FrameError("服务端只处理 REQ")
            if f.path not in self.handlers:
                self.scheduled.append((now, f.req_id, f.path,
                                       STATUS_UNKNOWN_PATH, b""))
                continue
            fn, delay = self.handlers[f.path]
            try:
                status, payload = 0, fn(f.payload)
            except Exception as e:      # 跨进程不传异常栈,只传码
                status, payload = getattr(e, "code", 5), b""
            self.scheduled.append((now + delay, f.req_id, f.path,
                                   status, payload))
        self.scheduled.sort(key=lambda s: (s[0], s[1]))
        while self.scheduled and self.scheduled[0][0] <= now:
            due, rid, path, status, payload = self.scheduled.pop(0)
            self.out.append(encode_frame(
                Frame(MSG_RESP, rid, path, payload, status)))

    def emit(self, now):
        while self.scheduled and self.scheduled[0][0] <= now:
            due, rid, path, status, payload = self.scheduled.pop(0)
            self.out.append(encode_frame(
                Frame(MSG_RESP, rid, path, payload, status)))


def check(label, cond, detail=""):
    assert cond, "%s %s" % (label, detail)
    print("[ok] %s" % label)


def main():
    # ---- 1. 帧编解码 ----
    f = Frame(MSG_RESP, 42, "/ZoneService/KickPlayer", b"bye", 0)
    raw = encode_frame(f)
    d = decode_frame(raw)
    check("帧往返:类型/req_id/路径/状态/载荷",
          (d.msg_type, d.req_id, d.path, d.status, d.payload)
          == (MSG_RESP, 42, "/ZoneService/KickPlayer", 0, b"bye"))
    check("长度前缀为 4 字节大端", raw[:4] == struct.pack(">I", len(raw) - 4))
    for label, bad in (("帧长不符", raw[:-1]),
                       ("未知类型", b"\x00\x00\x00\x03\x07\x00\x00\x00\x00\x00")):
        try:
            decode_frame(bad)
            check(label + " 应报错", False)
        except FrameError:
            check(label + " 报错", True)

    # ---- 2. 乱序响应按 req_id 正确关联 ----
    zone_to_global, global_to_zone = deque(), deque()
    cli = RpcClient(zone_to_global)
    srv = RpcServer(global_to_zone)
    srv.register("/Global/QueryRank", lambda p: b"rank-1", delay_ms=30)
    srv.register("/Global/QueryMail", lambda p: b"mail-7", delay_ms=5)
    srv.register("/Global/Ping", lambda p: b"pong", delay_ms=0)

    now = 0
    a = cli.call("/Global/QueryRank", timeout_ms=200)   # 慢
    b = cli.call("/Global/QueryMail", timeout_ms=200)
    c = cli.call("/Global/Ping", timeout_ms=200)
    srv.pump(zone_to_global, now)                       # 入队三响应 due 30/5/0
    srv.emit(now)
    while global_to_zone:
        cli.on_frame(global_to_zone.popleft())
    check("乱序到达仍正确关联(Ping 最先回)",
          cli.pending[c].result == b"pong"
          and cli.pending[a].result is None)
    srv.emit(now + 30)                                  # 30ms 后慢的才回
    while global_to_zone:
        cli.on_frame(global_to_zone.popleft())
    check("三个并发调用全部配对成功",
          cli.pending[a].result == b"rank-1"
          and cli.pending[b].result == b"mail-7"
          and cli.pending[c].result == b"pong")

    # ---- 3. 超时 + 晚到响应丢弃 ----
    d = cli.call("/Global/SlowEcho", b"x", timeout_ms=100)
    srv.register("/Global/SlowEcho", lambda p: p, delay_ms=200)
    srv.pump(zone_to_global, now)
    cli.expire(now + 110)
    check("110ms 判 TIMEOUT", cli.pending[d].error == STATUS_TIMEOUT)
    srv.emit(now + 200)                                 # 200ms 响应才到
    late = cli.on_frame(global_to_zone.popleft())
    check("晚到响应被静默丢弃", late == "dropped")

    # ---- 4. 错误传播(状态码)与未知路径 ----
    def boom(payload):
        class BizError(Exception):
            code = 7
        raise BizError()
    srv.register("/Global/Boom", boom, delay_ms=0)
    e = cli.call("/Global/Boom", timeout_ms=100)
    u = cli.call("/Global/NoSuch", timeout_ms=100)
    srv.pump(zone_to_global, now)
    srv.emit(now)
    while global_to_zone:
        cli.on_frame(global_to_zone.popleft())
    check("业务错误按状态码 7 传播", cli.pending[e].error == 7)
    check("未知路径返回状态码 2", cli.pending[u].error == STATUS_UNKNOWN_PATH)

    # ---- 5. 单向推送 ----
    srv.push("/ZoneService/KickPlayer", b"uid=1001")
    f = cli.on_frame(global_to_zone.popleft())
    check("PUSH 无配对直达推送队列",
          f != "dropped" and len(cli.pushes) == 1
          and cli.pushes[0].path == "/ZoneService/KickPlayer"
          and cli.pushes[0].payload == b"uid=1001")

    # ---- 6. pending 上限(应用层流控) ----
    cli2 = RpcClient(deque())
    for _ in range(64):
        cli2.call("/X/Y", timeout_ms=1000)
    check("在途调用数受控(64)", len(cli2.pending) == 64)

    print("\n全部断言通过")


if __name__ == "__main__":
    main()
